[system-probe] Add HTTP monitoring userspace side

ISauve opened the pull request:
### What does this PR do?

Adds the userspace side of HTTP monitoring in the system probe. The added components are responsible for processing HTTP transactions coming from ebpf, storing http transaction information in the appropriate ConnectionStats objects, and encoding the http data in order to be sent to the process agent & then along to the backend.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan
In the agent VM:
1) Enable HTTP monitoring in `system_probe.yaml`
    ```
    network_config:
          enable_http_monitoring: true
    ```
2) Build & run the system probe
    ```
    inv -e system-probe.build --bundle-ebpf && sudo ./bin/system-probe/system-probe
    ```
3) Make some http requests (must be http, not https)
     - a good endpoint to make requests to is `http://httpbin.org/status/{code}`, which generates a response with the given code
4) Fetch the connection stats from the system probe
    ```
    curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client\_id=1 | jq .
    ```
5) Find the connection stats corresponding to the http requests you made & verify that they were recorded in the "httpStatsByPath" field

ISauve committed:
Http monitoring userspace side

p-lambert commented on pkg/network/http/http_statkeeper.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+)
+
+type Key struct {
+	SourceIP   util.Address
+	DestIP     util.Address
+	SourcePort uint16
+	DestPort   uint16
+}
```
Just something to think about: For HTTP requests without keep alives each (src,dst) will map to a single request/path

ISauve replied:
Very good point - I ended up taking `SourcePort` out of the key for this reason

p-lambert commented on pkg/network/http/http_statkeeper.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+type Key struct {
+	SourceIP   util.Address
```
It might be worth it indexing the data using the high/low int64 representation (and doing conversions when you fetch the data), but let's profile first :)

p-lambert commented on pkg/network/http/monitor.go:
```diff
@@ -29,20 +28,11 @@ type Monitor struct {
 }
 
 // NewMonitor returns a new Monitor instance
-func NewMonitor(procRoot string, m *manager.Manager, h *ddebpf.PerfHandler) (*Monitor, error) {
-	filter, _ := m.GetProbe(manager.ProbeIdentificationPair{Section: string(probes.SocketHTTPFilter)})
-	if filter == nil {
-		return nil, fmt.Errorf("error retrieving socket filter")
-	}
-
-	closeFilterFn, err := network.HeadlessSocketFilter(procRoot, filter)
-	if err != nil {
-		return nil, fmt.Errorf("error enabling HTTP traffic inspection: %s", err)
-	}
-
```
Curious, why did you change this?

ISauve replied:
It was because there was an import cycle between the network & http packages - but this doesn't fix the tests, so I removed this change and instead made a new `filter` package to put `HeadlessSocketFilter` in

p-lambert reviewed (commented):
Gave a first pass on this. Overall I think it looks great!
I left a few comments mostly regarding performance concerns/optimizations that we'll likely need to consider at some point, but let's profile first :)

ISauve committed:
Continue work

- Update sketches-go dependency
- Switch quantile library
- Move ResponseStatus to http package
- Fix null pointer error
- Update agent-payload dependency
- Other bug fixes

ISauve committed:
Update bytecode

ISauve committed:
Restructure & update agent payload dependency

ISauve committed:
Update agent-payload dependency & modify imports accordingly

ISauve committed:
Create filter package

ISauve committed:
Add handler field back to http monitor

ISauve committed:
Add simple path cleaning

ISauve commented:
Note: currently this is blocked on https://github.com/DataDog/agent-payload/pull/78

sgnn7 commented on go.sum:
```diff
@@ -85,6 +85,25 @@ github.com/DATA-DOG/go-sqlmock v1.3.3/go.mod h1:f/Ixk793poVmq4qj/V1dPUg2JEAKC73Q
+github.com/DataDog/agent-payload v4.53.1-0.20210114152954-be676bf33059+incompatible h1:RbpQgkQIFyCRi3RMj5PCNmbb/g0u7AW2fDyKltOYWEs=
+github.com/DataDog/agent-payload v4.53.1-0.20210114152954-be676bf33059+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
+github.com/DataDog/agent-payload v4.53.1-0.20210114154107-05ad7aa9158d+incompatible h1:K62xDSk0vuDe64XpB1UdGJlT0X4YfrYyyBVYflqWnUM=
+github.com/DataDog/agent-payload v4.53.1-0.20210114154107-05ad7aa9158d+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
```
A small nit but I think the `.sum` file only needs one version's hashes so I think you can delete the v4.49 and 4.50.1 hashes from this file

sgnn7 commented on pkg/ebpf/bytecode/bindata/bindataOffsetguessO.go:
```diff
@@ -28,7 +28,7 @@ func bindataOffsetguessO() (*asset, error) {
 		size:        12360,
 		md5checksum: "",
-		mode:        os.FileMode(420),
+		mode:        os.FileMode(436),
```
Tiny nit: I don't know if these files are generated but it would be helpful to have a comment about what all these mode values are in octal right next to their use for easier visual parsing:
```suggestion
		mode:        os.FileMode(436),  // Octal: 0664
```

ISauve replied:
They are generated, so unfortunately that would get overwritten

ISauve committed:
Use updated agent-payload model

ISauve committed:
Clean up go.sum

ISauve committed:
Make HTTP stats 1:1 with network connection stats

p-lambert commented on pkg/network/http/http_statkeeper.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+type Key struct {
+	SourceIP util.Address
+	DestIP   util.Address
+	DestPort uint16
```
As mentioned in our chat yesterday, if you disregard SourcePort you're in effect losing container resolution

ISauve replied:
oops yes I did make that change but just forget to add it as part of this commit 😬   it's there now!

ISauve committed:
Add tests

ISauve committed:
Add debug logging

ISauve committed:
Address PR feedback

ISauve committed:
Remove +build linux_bpf from http package files

ISauve commented on pkg/network/http/model.go:
```diff
@@ -82,6 +90,36 @@ func (tx *httpTX) Method() string {
+}
+
+func (tx *httpTX) RequestLatency() time.Duration {
+	return time.Duration(int64((tx.response_last_seen-tx.request_started)/(1000000))) * time.Millisecond
```
Yeah this was an error - I wanted to convert from ns to ms and thought I needed `* time.Millisecond` at the end for time.Duration to know the value I was giving it was in ms 😅  I ended up switching to use a float representation anyways though

ISauve commented on pkg/network/http/request_stats.go:
```diff
@@ -0,0 +1,57 @@
+// RequestStats stores stats for HTTP requests to a particular path, organized by the class
+// of the response code (1XX, 2XX, 3XX, 4XX, 5XX)
+type RequestStats [5]struct {
+	count     int
```
It does, but technically every time we add a value to the sketch it's possible for the sketch to not accept that value (ie if the value is outside the range that is tracked by the sketch), so I figured keeping a separate count which we know captures 100% of processed transactions is better

p-lambert replied:
I guess we can keep it for now then! Could you just please add a small comment explaining your motivation?

ISauve replied:
Definitely :)

ISauve commented on pkg/network/http/request_stats.go:
```diff
@@ -0,0 +1,57 @@
+	r[i].count++
+
+	if r[i].latencies == nil {
+		relativeAccuracy := 0.01
```
👍  good idea!

ISauve committed:
Generate licenses

ISauve committed:
Merge branch 'master' into isabelle.sauve/http-monitoring-userspace

ISauve committed:
Fix build issues

- Add +build linux_bpf tag back to files
- Move http.Key to httpStats
- Fix go vet errors
- Update agent-payload dependency
- Use PacketSource interface in dns_snooper.go & rename PacketSource -> AFPacketSource
- Fix state test

ISauve committed:
Add comment explaining RequestStats.count field

ISauve committed:
Encode latencies as byte arrays

p-lambert reviewed (approved):
Awesome work!
The only thing that this PR is missing in my opinion is a proper integration test.
I think it would be great if you could add one, or modify `TestEnableHTTPMonitoring` to do a http request, call `GetActiveConnections` and see that you have the associated HTTP stats.

ISauve commented on pkg/network/encoding/encoding_test.go:
```diff
@@ -89,6 +95,33 @@ func TestSerialization(t *testing.T) {
 					},
 				},
+
+				HttpStatsByPath: map[string]*model.HTTPStats{
+					"/testpath": {
+						StatsByResponseStatus: []*model.HTTPStats_Data{
+							{
```
Great idea 👍

ISauve committed:
Use proto.Marshal to do latency serialization

ISauve committed:
Add integration test

Fix encoding test

ISauve committed:
Merge branch 'master' into isabelle.sauve/http-monitoring-userspace

ISauve committed:
Update agent payload dependency to release v4.55.0
